File: mobile/ios/OpenU60/Features/RouterSettings/APN/APNViewModel.swift
import SwiftUI

@Observable
@MainActor
final class APNViewModel {
    var config: APNConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false
    var showFormSheet: Bool = false

    // Form state
    var formProfile: APNProfile = .empty
    var editingProfile: APNProfile?  // nil = adding, non-nil = editing
    var setAsDefault: Bool = false

    private let client: AgentClient
    private let authManager: AuthManager

    var isEditing: Bool { editingProfile != nil }

    var activeAPNName: String? {
        let active = config.profiles.first(where: { $0.active })
            ?? config.autoProfiles.first(where: { $0.active })
        guard let active else { return nil }
        return active.name.isEmpty ? active.apn : active.name
    }

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let modeData = try await client.getJSON("/api/router/apn/mode")
            let mode = APNParser.parseMode(modeData)

            let manuData = try await client.getJSON("/api/router/apn/profiles")
            let profiles = APNParser.parseProfiles(manuData)

            let autoData = try await client.getJSON("/api/router/apn/auto-profiles")
            let autoProfiles = APNParser.parseProfiles(autoData)

            config = APNConfig(mode: mode, profiles: profiles, autoProfiles: autoProfiles)
        } catch {
            showMessage("加载 APN 失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func setMode(manual: Bool) async {
        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/apn/mode", body: ["apn_mode": manual ? 1 : 0])
            showMessage(manual ? "已切换为手动 APN 模式" : "已切换为自动 APN 模式", isError: false)
            config = APNConfig(
                mode: manual ? "1" : "0",
                profiles: config.profiles,
                autoProfiles: config.autoProfiles
            )
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    // MARK: - Form Actions

    func startAdd() {
        editingProfile = nil
        formProfile = .empty
        setAsDefault = false
        showFormSheet = true
    }

    func startEdit(_ profile: APNProfile) {
        editingProfile = profile
        formProfile = profile
        setAsDefault = profile.active
        showFormSheet = true
    }

    func saveAPN() async {
        guard !formProfile.name.isEmpty, !formProfile.apn.isEmpty else {
            showMessage("名称和 APN 不能为空", isError: true)
            return
        }

        let isDuplicate = config.profiles.contains { p in
            p.name == formProfile.name && p.id != (editingProfile?.id ?? "")
        }
        if isDuplicate {
            showMessage("已存在同名 APN", isError: true)
            return
        }

        if isEditing {
            await editAPN()
        } else {
            await addAPN()
        }
    }

    private func addAPN() async {
        isLoading = true

        let name = formProfile.name
        let apn = formProfile.apn
        let shouldSetDefault = setAsDefault

        do {
            let _ = try await client.postJSON("/api/router/apn/profiles", body: [
                "profilename": name,
                "wanapn": apn,
                "pdpType": formProfile.pdpType,
                "pppAuthMode": formProfile.authMode,
                "username": formProfile.username,
                "password": formProfile.password
            ])

            if shouldSetDefault {
                await refresh()
                if let newProfile = config.profiles.first(where: { $0.name == name && $0.apn == apn }) {
                    await activateAPN(newProfile)
                }
            }

            formProfile = .empty
            showFormSheet = false
            showMessage("APN 已添加", isError: false)
            await refresh()
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func editAPN() async {
        guard let editing = editingProfile else { return }

        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/apn/profiles", body: [
                "profileId": editing.id,
                "profilename": formProfile.name,
                "wanapn": formProfile.apn,
                "pdpType": formProfile.pdpType,
                "pppAuthMode": formProfile.authMode,
                "username": formProfile.username,
                "password": formProfile.password
            ])

            if setAsDefault && !editing.active {
                await activateAPN(editing)
            }

            editingProfile = nil
            showFormSheet = false
            showMessage("APN 已更新", isError: false)
            await refresh()
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func deleteAPN(_ profile: APNProfile) async {
        if profile.active {
            showMessage("无法删除当前正在使用的 APN", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.postJSON("/api/router/apn/profiles/delete", body: ["profileId": profile.id])
            showMessage("APN 已删除", isError: false)
            config.profiles.removeAll { $0.id == profile.id }
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func activateAPN(_ profile: APNProfile) async {
        isLoading = true

        do {
            let _ = try await client.postJSON("/api/router/apn/profiles/activate", body: ["profileId": profile.id])
            showMessage("APN 已启用", isError: false)
            config.profiles = config.profiles.map { p in
                var updated = p
                updated.active = p.id == profile.id
                return updated
            }
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/apn/APNViewModel.kt
package com.openu60.feature.router.apn

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.APNConfig
import com.openu60.core.model.APNParser
import com.openu60.core.model.APNProfile
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class APNState(
    val config: APNConfig = APNConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
    val showForm: Boolean = false,
    val editingProfile: APNProfile? = null,
)

@HiltViewModel
class APNViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(APNState())
    val state: StateFlow<APNState> = _state.asStateFlow()

    fun showAddForm() {
        _state.value = _state.value.copy(showForm = true, editingProfile = APNProfile.empty)
    }

    fun showEditForm(profile: APNProfile) {
        _state.value = _state.value.copy(showForm = true, editingProfile = profile)
    }

    fun hideForm() {
        _state.value = _state.value.copy(showForm = false, editingProfile = null)
    }

    fun updateEditingProfile(profile: APNProfile) {
        _state.value = _state.value.copy(editingProfile = profile)
    }

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val modeData = agentClient.getJSON("/api/router/apn/mode")
                val profilesData = agentClient.getJSON("/api/router/apn/profiles")
                val autoProfilesData = agentClient.getJSON("/api/router/apn/auto-profiles")
                val mode = APNParser.parseMode(modeData)
                val profiles = APNParser.parseProfiles(profilesData)
                val autoProfiles = APNParser.parseProfiles(autoProfilesData)
                _state.value = _state.value.copy(
                    config = APNConfig(mode = mode, profiles = profiles, autoProfiles = autoProfiles),
                    isLoading = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun setMode(mode: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/router/apn/mode", mapOf("apn_mode" to (mode.toIntOrNull() ?: 0)))
                _state.value = _state.value.copy(
                    config = _state.value.config.copy(mode = mode),
                    isLoading = false,
                    message = if (mode == "1") "已切换为手动 APN 模式" else "已切换为自动 APN 模式",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) setMode(mode) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun saveProfile() {
        val profile = _state.value.editingProfile ?: return
        val existingNames = _state.value.config.profiles
            .filter { it.id != profile.id }
            .map { it.name.lowercase() }
        if (profile.name.lowercase() in existingNames) {
            _state.value = _state.value.copy(message = "配置名称已存在", messageIsError = true)
            return
        }

        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val params = mapOf(
                    "profileId" to profile.id,
                    "profilename" to profile.name,
                    "wanapn" to profile.apn,
                    "pdpType" to profile.pdpType,
                    "pppAuthMode" to profile.authMode,
                    "username" to profile.username,
                    "password" to profile.password,
                )
                if (profile.id.isNotBlank()) {
                    agentClient.putJSON("/api/router/apn/profiles", params)
                } else {
                    agentClient.postJSON("/api/router/apn/profiles", params)
                }
                _state.value = _state.value.copy(showForm = false, editingProfile = null, isLoading = false, message = "配置已保存", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) saveProfile() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun deleteProfile(id: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/router/apn/profiles/delete", mapOf("profileId" to id))
                _state.value = _state.value.copy(isLoading = false, message = "配置已删除", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) deleteProfile(id) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun activateProfile(id: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/router/apn/profiles/activate", mapOf("profileId" to id))
                _state.value = _state.value.copy(isLoading = false, message = "配置已启用", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) activateProfile(id) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "未知错误", messageIsError = true)
    }
}
